File: LinkedLists/Java/Add2Nums.java
package LinkedLists.Java;

public class Add2Nums {
    private static class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    private static Node convertArrayToLL(int[] arr) {
        if (arr.length == 0) {
            return null;
        }
        Node head = new Node(arr[0]);
        Node mover = head; // This will be used to traverse and add new nodes
        for (int i = 1; i < arr.length; i++) {
            Node temp = new Node(arr[i]); // Create a new node with the current array element
            mover.next = temp; // Link the current node to the new node
            mover = temp; // Move the mover pointer to the new node
        }
        return head;
    }

    private static void printLinkedList(Node head) {
        while (head != null) {
            System.out.print(head.data + "->");
            head = head.next;
        }
        System.out.println("null");
    }

    public static Node addTwoNumbers(Node l1, Node l2) {
        // Time Complexity: O(max(N, M)) where N and M are the lengths of the two linked lists
        // Space Complexity: O(max(N, M)) for the result linked list
        Node dummyHead = new Node(0); // This node is to traverse and build the result linked list
        Node p = l1, q = l2, current = dummyHead;
        int carry = 0;

        while (p != null || q != null) {
            int x = (p != null) ? p.data : 0;
            int y = (q != null) ? q.data : 0;
            int sum = carry + x + y;
            carry = sum / 10;
            current.next = new Node(sum % 10); // This is the node to be added to the result linked list
            current = current.next;

            if (p != null) p = p.next;
            if (q != null) q = q.next;
        }
        if (carry > 0) {
            current.next = new Node(carry);
        }
        return dummyHead.next;
    }

    public static void main(String[] args) {
        Node head1 = convertArrayToLL(new int[]{9, 9, 3});
        Node head2 = convertArrayToLL(new int[]{7, 8, 4, 9});
        Node result = addTwoNumbers(head1, head2);
        System.out.print("Result Linked List is: ");
        printLinkedList(result);
    }
}

File: LinkedLists/Python/add2Nums.py
class Node:
    def __init__(self,data1,next1 = None):
        self.data = data1
        self.next = next1
        
def convertarr2LL(arr):
    head = Node(arr[0])
    mover = head
    
    for i in range(1,len(arr)):
        tmp = Node(arr[i])
        mover.next = tmp
        mover = tmp
        
    return head
    
def add(head1,head2):
    t1,t2 = head1,head2
    dummy = Node(-1) # Place holder to make the retrievel easy
    curr = dummy # Pointer to traverse
    carry = 0
    
    # Traverse across both the lists
    while t1 or t2:
        res = carry
        # Add the sum of each lists into result
        if t1:
            res = res + t1.data
        if t2:
            res = res + t2.data
        
        # Compute the result for each digit
        newNode = Node(res%10) # This will yield the reminder of sum
        carry = res//10 # This yields the carry over
        
        curr.next = newNode
        curr = curr.next
        
        if t1:
            t1 = t1.next
        if t2:
            t2 = t2.next
        
    if carry:
        newNode = Node(carry)
        curr.next = newNode # As current is at the tail just point it to the new node
        
    return dummy.next
        
def printLL(head):
    while head:
        print(head.data,end=' ')
        head = head.next
    print()
    
def main():
    arr = [3,5]
    arr2 = [4,8,9,9]
    res1 = convertarr2LL(arr)
    res2 = convertarr2LL(arr2)
    res = add(res1,res2)
    print("The linked list is: ",end='')
    printLL(res)
    
if __name__ == "__main__":
    main()